pkg/builder/naive_build_directory.go

```diff
@@ -56,36 +60,67 @@ func (d *naiveBuildDirectory) mergeDirectoryContents(ctx context.Context, digest
 	}
 
-	// Create children.
+	actionGroup := re_util.ActionGroup{}
+
 	for _, file := range directory.Files {
-		childComponents := append(components, file.Name)
+		childComponents := append([]string(nil), components...) // We need a copy of our slice to pass to go routine.
+		childComponents = append(childComponents, file.Name)
 		childDigest, err := digest.NewDerivedDigest(file.Digest)
 		if err != nil {
 			return util.StatusWrapf(err, "Failed to extract digest for input file %#v", path.Join(childComponents...))
 		}
-		if err := d.contentAddressableStorage.GetFile(ctx, childDigest, inputDirectory, file.Name, file.IsExecutable); err != nil {
-			return util.StatusWrapf(err, "Failed to obtain input file %#v", path.Join(childComponents...))
-		}
+		// Extract values out of object so we don't end up with stale objects when we pass into go routine.
+		fileName := file.Name
+		fileIsExecutable := file.IsExecutable
+
+		actionGroup = append(actionGroup, func(actionContext *re_util.ActionContext) {
+			if err := d.contentAddressableStorage.GetFile(actionContext.Ctx(), childDigest, inputDirectory, fileName, fileIsExecutable); err != nil {
+				actionContext.Done(util.StatusWrapf(err, "Failed to obtain input file %#v", path.Join(childComponents...)))
+				return
+			}
+			actionContext.Done(nil)
+		})
 	}
+
 	for _, directory := range directory.Directories {
-		childComponents := append(components, directory.Name)
+		childComponents := append([]string(nil), components...) // We need a copy of our slice to pass to go routine.
+		childComponents = append(childComponents, directory.Name)
 		childDigest, err := digest.NewDerivedDigest(directory.Digest)
 		if err != nil {
 			return util.StatusWrapf(err, "Failed to extract digest for input directory %#v", path.Join(childComponents...))
 		}
 		if err := inputDirectory.Mkdir(directory.Name, 0777); err != nil {
 			return util.StatusWrapf(err, "Failed to create input directory %#v", path.Join(childComponents...))
 		}
-		childDirectory, err := inputDirectory.EnterDirectory(directory.Name)
-		if err != nil {
-			return util.StatusWrapf(err, "Failed to enter input directory %#v", path.Join(childComponents...))
-		}
-		err = d.mergeDirectoryContents(ctx, childDigest, childDirectory, childComponents)
-		childDirectory.Close()
-		if err != nil {
-			return err
-		}
+		directoryName := directory.Name // Ensure we don't grab the reference to an object.
+		actionGroup = append(actionGroup, func(actionContext *re_util.ActionContext) {
+			childDirectory, err := inputDirectory.EnterDirectory(directoryName)
+			if err != nil {
+				actionContext.Done(util.StatusWrapf(err, "Failed to enter input directory %#v", path.Join(childComponents...)))
+				return
+			}
+			// We don't want to close the folder's handle until our children are done, however we also don't
+			// want to hold a slot of the SharedActionQueue (or we might deadlock). Since 'mergeDirectoryContents'
```

Comment:
The deadlock we would otherwise run into is in my opinion a symptom of us trying to introduce too much parallelism. Imagine this input root that has an awful number of directories. Millions and millions of directories, each nested into each other. Just ignore any of the files that are inside of them (if any). Without the defer logic in here, the queue would deadlock, for the reason that all of the slots are filled with directory processing actions, all trying to yield even more actions that process directories.

The defer logic/goroutine breakup you added prevents the deadlock, but the question is what we lose as a result of that. The answer to that is that there's no upper bound on the concurrent number of directories we're processing. This means there's also no upper bound on memory use, directory file descriptors held, etc. etc. etc.

Maybe we should take a step back and see how we can add parallelism without unbounded resource usage. What if you were to walk the directory hierarchy synchronously (like we already do right now), but at least add parallelism downloading/hardlinking files? Maybe that already brings performance close to what you desire without complicating the design too much.

Reply:
Hmmm, I'm not thrilled about the idea of restricting walking directories to be synchronous. I suppose it wouldn't be too bad to do them on a per-level basis, but I think this would still have the open folder problem since one folder could have hundreds of folders inside it, so no matter how you look at it, we will always have at least MAX_SLOTS folder descriptors open. There's also the problem of complexity this would induce. This PR is already adding a bit of complexity as it is which I'm not happy with at all.

Maybe we could do this in passes and instead of trying to place the files/folders immediately we place the files into a flat folder with hashes/indexes then once all of them are available we could materialize the folder tree by making each folder and moving each file into the retrospective directory?

There's also a few things we could do to significantly reduce the chance of hitting our limit (but not removing entirely), like only allowing one thread to process the "file" for-loop at a time. This way statistically we'd have most of our files clumped together along with a refcounter to know when it is safe to close the file descriptor of the parent folder.

I personally prefer the first option, but will mean we'd need to either abuse the `cache` folder structure or create another folder for this temp stuff (which I'd strongly object towards).

Reply:
> [...] we will always have at least MAX_SLOTS folder descriptors open. [...]

Wait. There's a subtle difference here.

With the solution that I'm proposing, the number of file descriptors that are open will be `O(k*d + n)`, where `k` is the concurrency of the worker, `d` is the maximum depth of an input root and `n` is the maximum slots for parallel uploading.

With the change that you're proposing, there is absolutely no bound at all. It's `O(∞)` (if that's even a thing).

Sure, there is a quadratic component in what I'm proposing and it's true what you state: it's at least MAX_SLOTS. Still, that's better than having absolutely no bound whatsoever.

Reply:
I don't think so, the complexity should be only `O(max(MAX_SLOTS, FOLDER_DEPTH))`.

It would follow something like:
* Create a temp folder
* Download each file into this directory (no sub-folders) with the key being the file's hash.
* After all files are downloaded walk the digest tree again and build the directory tree and move/hardlink the files to the new directory to the proper location.
* Delete temp folder

[we could use the same folder instead of creating a temp folder, but for illustration reasons I kept them separate].

FYI, we are not having issues with uploading from the worker [yet] but are having issues downloading to the worker. In our case it is because we have a folder that tens of thousands of actions use to compile (python) and this digest/folder has ~10k files inside it. Once a worker downloads it it is cached, but since we are using spot instances our workers die frequently, so we are trying to optimize this specific use case.